cmd/flux/diff_artifact.go

```diff
@@ -98,15 +139,308 @@ func diffArtifactCmdRun(cmd *cobra.Command, args []string) error {
+		}
+	} else {
+		fh, err := os.Open(path)
+		if err != nil {
+			cleanup()
+			return "", nil, fmt.Errorf("os.Open(%q): %w", path, err)
+		}
+		defer fh.Close()
+
+		name := filepath.Join(tmpDir, filepath.Base(path))
+		if err := copyFile(fh, name); err != nil {
+			cleanup()
+			return "", nil, fmt.Errorf("os.Open(%q): %w", path, err)
+		}
+	}
+
+	return tmpDir, cleanup, nil
+}
+
+// extractTo extracts the .tar.gz / .tgz archive at archivePath into the destDir directory.
+func extractTo(archivePath, destDir string) error {
+	archiveFH, err := os.Open(archivePath)
+	if err != nil {
 		return err
 	}
 
-	logger.Successf("no changes detected")
+	gzipReader, err := gzip.NewReader(archiveFH)
+	if err != nil {
+		return fmt.Errorf("gzip.NewREader(): %w", err)
+	}
+
+	tarReader := tar.NewReader(gzipReader)
+
+	for {
+		header, err := tarReader.Next()
+		if errors.Is(err, io.EOF) {
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("tarReader.Next(): %w", err)
+		}
+
+		switch header.Typeflag {
+		case tar.TypeDir:
+			dir := filepath.Join(destDir, header.Name)
+			if err := os.Mkdir(dir, 0755); err != nil {
+				return fmt.Errorf("os.Mkdir(%q): %w", dir, err)
+			}
+
+		case tar.TypeReg:
+			name := filepath.Join(destDir, header.Name)
+			if err := copyFile(tarReader, name); err != nil {
+				return fmt.Errorf("extracting %q: %w", header.Name, err)
+			}
+
+		default:
+			logger.Warningf("unsupported tar type: %v", header.Typeflag)
+		}
+	}
+
+	return nil
+}
+
+func copyFile(from io.Reader, to string) error {
+	fh, err := os.Create(to)
+	if err != nil {
+		return fmt.Errorf("os.Create(%q): %w", to, err)
+	}
+	defer fh.Close()
+
+	if _, err := io.Copy(fh, from); err != nil {
+		return fmt.Errorf("io.Copy(%q): %w", to, err)
+	}
+
 	return nil
 }
+
+type differ interface {
+	// Diff compares the two local directories "to" and "from" and returns their differences, or an empty string if they are equal.
+	Diff(ctx context.Context, from, to string) (string, error)
+}
+
+// externalDiffCommand implements the differ interface using an external diff command.
+type externalDiffCommand struct {
+	name  string
+	flags []string
+}
+
+// externalDiffVar is the environment variable users can use to overwrite the external diff command.
+const externalDiffVar = "FLUX_EXTERNAL_DIFF"
+
+// mustExternalDiff initializes an externalDiffCommand using the externalDiffVar environment variable.
+func mustExternalDiff() externalDiffCommand {
+	cmdline := os.Getenv(externalDiffVar)
+	if cmdline == "" {
+		cmdline = "diff -ur"
+	}
+
+	args, err := shlex.Split(cmdline)
+	if err != nil {
+		panic(fmt.Sprintf("shlex.Split(%q): %v", cmdline, err))
+	}
+
+	return externalDiffCommand{
+		name:  args[0],
+		flags: args[1:],
+	}
+}
+
+func (c externalDiffCommand) Diff(ctx context.Context, fromDir, toDir string) (string, error) {
+	var args []string
+
+	args = append(args, c.flags...)
+	args = append(args, fromDir, toDir)
+
+	cmd := exec.CommandContext(ctx, c.name, args...)
```

Comment:
✅ done – the `--differ` flag now takes one of three arguments:

1. `dyff`: Use the `dyff` package to generate a semantic YAML diff.
2. `external`: execute the command in the `FLUX_EXTERNAL_DIFF` environment variable. Using `--differ external` without setting `FLUX_EXTERNAL_DIFF` is now an error, there is no default command being executed. That means flux makes no assumptions about the system is runs on.
3. `unified`: uses [github.com/hexops/gotextdiff](https://pkg.go.dev/github.com/hexops/gotextdiff) to generate a unified diff without invoking external commands.

   I looked into unified diffs and Golang quite a bit more than I wanted to — the situation is surprisingly bad. I opted to use [github.com/hexops/gotextdiff](https://pkg.go.dev/github.com/hexops/gotextdiff), which is a clone of one of the internal-only packages by the Go authors. The repo is marked as archived as of earlier this year, but overall this seems like the least bad option.

> In the past, we have taken other such code from upstream Go and adapted it for our needs

As a maintainer myself, I'd work hard to avoid that. Copying and adapting code rather than referencing it means that all the maintenance overhead is on the Flux maintainers, which can snowball quickly.